rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+way around this as if we were to exclude the results marked `mature` from the
+origin result IDs in the final set subtraction, we'd end up with the following
+situation:
+
+```
+origin_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+  {id: 2, mature: true}, # no sensitive text
+  {id: 3, mature: false}, # has sensitive text
+  {id: 4, mature: true}, # has sensitive text
+]
+
+filtered_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+]
+
+origin_index_less_mature_result_ids = [r for r in origin_index results if not r["mature"]]
+# => [1, 3]
+
+origin_index_less_mature_result_ids - filtered_index_result_ids
+# => [1, 3] - [1] = [3]
+```
+
+Notice how we still have a problem: we've excluded the result that has sensitive
+text _and_ is marked mature (result 4).
+
+It's tempting to think we could move the mature filtering to _after_ when we've
+derived the list of sensitive result IDs. However, the problem will remain the
+same: we will exclude results that are both marked mature and have sensitive
+textual content:
+
+```
+origin_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+  {id: 2, mature: true}, # no sensitive text
+  {id: 3, mature: false}, # has sensitive text
+  {id: 4, mature: true}, # has sensitive text
+]
+
+filtered_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+]
+
+sensitive_result_ids = origin_index_result_ids - filtered_index_result_ids
+# => [1, 2, 3, 4] - [1] = [2, 3, 4]
+
+sensitive_results = [r for r in origin_index_results if r["_id"] in sensitive_result_ids and not r["mature"]]
+# => [3]
+```
+
+Result 4 is still excluded because we can't tell in the final operation whether
+result 4 was not in the set of IDs from the filtered index _only_ because it is
+marked mature or if it also has sensitive textual content.
+
+The only way I can think of to solve this is to have an index solely of
+documents that _do_ have sensitive textual content and querying it to derive the
+final list of sensitive results without needing to rely on the diversely
+filtered index. However, this adds a big re-indexing burden (yet another
+additional index) that can be bypassed altogether if we just exclude mature
+results from the filtered index to begin with.
+
+## Wrap up
+
+To wrap up the implementation, once the final step has been completed and
+results are filtered or correctly marked as sensitive, we will promote the
+`include_sensitive_results` parameter by removing the `unstable` designation.
+Additionally, we will deprecate the `mature` parameter by removing it from the
+documentation. To maintain API version backwards compatibility, it should
+continue to operate as an undocumented alias for the `include_sensitive_results`
+parameter.
+
+## Feature flags
+
+Because the new functionality will all be either internal (ingestion server and
+catalogue) or behind an unstable query parameter, we do not need any new feature
+flags to manage the rollout of this work.
```

Comment:
Since `include_sensitive_results` is by default False, even though the actual query parameter is unstable won't this be immediately rolled out in production once the first data refresh including the new index changes completes?

Do we have a planned way to roll back/query from the original index in case of some problem with the filtered index in production? Is that necessary?

Reply:
> Do we have a planned way to roll back/query from the original index in case of some problem with the filtered index in production? Is that necessary?

That's a good point, I hadn't considered the mismatch there. Let's add a Django setting for enabling the new query parameter.